### PR TITLE
Use port 0 InternalTestCluster nodes

### LINKED ISSUE #27762: [CI] MinimumMasterNodesIT tests failed periodically
Some tests in `MinimumMasterNodesIT` started failing recently.

1. `MinimumMasterNodesIT#testCanNotPublishWithoutMinMastNodes`: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.0+periodic/1038/console and https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+intake/993/console

2. `MinimumMasterNodesIT#testMultipleNodesShutdownNonMasterNodes` : https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+oracle-periodic/1642/console and https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.0+java9-periodic/1645/console

3. `MinimumMasterNodesIT#testCanNotBringClusterDown` : https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.0+aggressive-opts/482/console

4. `MinimumMasterNodesIT#testDynamicUpdateMinimumMasterNodes`: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+5.6+multijob-unix-compatibility/os=ubuntu/421/console and https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+release-tests/280/console

### LINKED ISSUE #27818: Master node not discovered in CI
_Note: These are three different test cases that fail but it appears to be the same cause, hence I just created one ticket._

Example build failure: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+periodic/5368/console

The [full build log](https://github.com/elastic/elasticsearch/files/1559329/build.log.zip) is attached.

failing tests:

```
org.elasticsearch.indices.state.RareClusterStateIT#testDelayedMappingPropagationOnPrimary
org.elasticsearch.cluster.routing.PrimaryAllocationIT#testDoNotAllowStaleReplicasToBePromotedToPrimary
org.elasticsearch.cluster.MinimumMasterNodesIT#testSimpleMinimumMasterNodes
```

Example failure message (more of them in the attached log):

```
ERROR   61.2s J2 | MinimumMasterNodesIT.testSimpleMinimumMasterNodes <<< FAILURES!
   > Throwable #1: MasterNotDiscoveredException[null]
   > 	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$4.onTimeout(TransportMasterNodeAction.java:230)
   > 	at org.elasticsearch.cluster.ClusterStateObserver$ContextPreservingListener.onTimeout(ClusterStateObserver.java:317)
   > 	at org.elasticsearch.cluster.ClusterStateObserver$ObserverClusterStateListener.onTimeout(ClusterStateObserver.java:244)
   > 	at org.elasticsearch.cluster.service.ClusterApplierService$NotifyTimeout.run(ClusterApplierService.java:581)
   > 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:566)
   > 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
   > 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
   > 	at java.lang.Thread.run(Thread.java:748)Throwable #2: MasterNotDiscoveredException[null]
   > 	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$4.onTimeout(TransportMasterNodeAction.java:230)
   > 	at org.elasticsearch.cluster.ClusterStateObserver$ContextPreservingListener.onTimeout(ClusterStateObserver.java:317)
   > 	at org.elasticsearch.cluster.ClusterStateObserver$ObserverClusterStateListener.onTimeout(ClusterStateObserver.java:244)
   > 	at org.elasticsearch.cluster.service.ClusterApplierService$NotifyTimeout.run(ClusterApplierService.java:581)
   > 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:566)
   > 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
   > 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
   > 	at java.lang.Thread.run(Thread.java:748)
```

before this failure we see in the log:

```
  1> [2017-12-14T08:01:25,025][WARN ][o.e.d.z.UnicastZenPing   ] [node_t1] [299] failed send ping to {#zen_unicast_127.0.0.1_3#}{QVh9ejEISw-Mkk7JqYIF1g}{127.0.0.1}{127.0.0.1:9623}
  1> java.lang.IllegalStateException: handshake failed, mismatched cluster name [Cluster [SUITE-CHILD_VM=[2]-CLUSTER_SEED=[-3582321890366079432]-HASH=[4DDF2E03F454A3]-cluster]] - {#zen_unicast_127.0.0.1_3#}{QVh9ejEISw-Mkk7JqYIF1g}{127.0.0.1}{127.0.0.1:9623}
```

Reproduction line:

```
REPRODUCE WITH: gradle :core:integTest \
  -Dtests.seed=3049C5F03325CE2F \
  -Dtests.class=org.elasticsearch.cluster.MinimumMasterNodesIT \
  -Dtests.method="testSimpleMinimumMasterNodes" \
  -Dtests.security.manager=true \
  -Dtests.locale=es-GT \
  -Dtests.timezone=America/Eirunepe

REPRODUCE WITH: gradle :core:integTest \
  -Dtests.seed=3049C5F03325CE2F \
  -Dtests.class=org.elasticsearch.cluster.routing.PrimaryAllocationIT \
  -Dtests.method="testDoNotAllowStaleReplicasToBePromotedToPrimary" \
  -Dtests.security.manager=true \
  -Dtests.locale=ar-TN \
  -Dtests.timezone=Australia/Adelaide

REPRODUCE WITH: gradle :core:integTest \
  -Dtests.seed=3049C5F03325CE2F \
  -Dtests.class=org.elasticsearch.indices.state.RareClusterStateIT \
  -Dtests.method="testDelayedMappingPropagationOnPrimary" \
  -Dtests.security.manager=true \
  -Dtests.locale=sk-SK \
  -Dtests.timezone=Asia/Kathmandu
```

reproduces locally: no (I ran a full build with `gradle :core:integTest -Dtests.seed=3049C5F03325CE2F -Dtests.security.manager=true -Dtests.locale=sk-SK -Dtests.timezone=Asia/Kathmandu`)

failure frequency:

* `RareClusterStateIT#testDelayedMappingPropagationOnPrimary`: 10 times within the last year (out of that 6 times since December 11)
* `PrimaryAllocationIT#testDoNotAllowStaleReplicasToBePromotedToPrimary`: 15 times within the last year (out of that 10 times since December 11)
* `MinimumMasterNodesIT#testSimpleMinimumMasterNodes`: 20 times within the last year (out of that 12 times since December 11)

### PR DESCRIPTION
**Pull request by bleskes**
We currently have a complicated port assignment scheme to make sure that the nodes span off by the internal test cluster will be assigned fixed port ranges that will also not collide between clusters. The port ranges need to be fixed in advance so that the nodes will be able to find each other via `UnicastZenPing`.

This approach worked well for the last few years but we are now at a point that our testing has grown beyond it and we exceed the 5 reusable ranges per JVM. This means that nodes are not always assigned the first 5 ports in their range which causes cluster formation issues. On top of that, most of the clusters that are span up don't even rely on `UnicastZenPing` but rather `MockZenPings` that uses in memory maps for discovery (with the down side that they are not influenced by network disruption simulations).

This PR changes `InternalTestCluster` to use port 0 as a fixed assignment. This will allow the OS to manage ports and will ensure we don't have collisions. For tests that need to simulate network disruptions (and thus can't use `MockZenPings`), a new `UnicastHostProvider` is introduced that is based on the current state of the test cluster. Since that is only resolved at run time, it is aware of the port assignments of the OS.

Closes #27818
Closes #27762

### CODE REVIEW

**Review by ywelsch — COMMENTED** [review]
I wonder if we should use the [file-based discovery](https://www.elastic.co/guide/en/elasticsearch/plugins/current/discovery-file.html) instead (and move that discovery method to core). The advantages would be:
- No need to use a mock, we can just use the real thing.
- We can use file-based discovery in our REST tests, which would also fix a similar issue to the current one here (see https://github.com/elastic/elasticsearch/pull/27344)

**Comment by bleskes** [comment]
> I wonder if we should use the file-based discovery instead (and move that discovery method to core).

That's a good idea. I'll explore that. Thanks!

**Comment by bleskes** [comment]
> > I wonder if we should use the file-based discovery instead (and move that discovery method to core).

> That's a good idea. I'll explore that. Thanks!

I turns out this is difficult to do because the test infra doesn't depend on plugins and we want to keep it like that. Folding into core would cause another set of discussions. I'll address your existing comments.

**Review by ywelsch — APPROVED** [review]
LGTM

File: test/framework/src/main/java/org/elasticsearch/test/discovery/TestZenDiscovery.java [hunk 1]
```diff
@@ -53,28 +59,65 @@
...
             return Collections.singletonList(USE_MOCK_PINGS);
         }
 
         @Override
         public Settings additionalSettings() {
-            return Settings.builder().put(DiscoveryModule.DISCOVERY_TYPE_SETTING.getKey(), "test-zen").build();
+            return Settings.builder()
+                .put(DiscoveryModule.DISCOVERY_TYPE_SETTING.getKey(), "test-zen")
+                .put(DiscoveryModule.DISCOVERY_HOSTS_PROVIDER_SETTING.getKey(), "test-zen")
+                .putList(DISCOVERY_ZEN_PING_UNICAST_HOSTS_SETTING.getKey())
```
[the remark this answers is not in the document]
Comment (bleskes) [reply test/framework/src/main/java/org/elasticsearch/test/discovery/TestZenDiscovery.java]: This is needed to completely disable the default addresses. I wanted to lock things down and avoid ugly messages in the logs.

**Comment by bleskes** [comment]
thx @ywelsch

**Pull request merged**